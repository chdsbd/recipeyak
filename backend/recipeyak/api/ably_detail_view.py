from __future__ import annotations

import asyncio

from ably import AblyRest
from rest_framework import status
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from recipeyak.api.base.request import AuthedRequest
from recipeyak.api.team_update_view import get_teams
from recipeyak.config import ABLY_API_KEY


async def get_token(user_id: str, team_ids: list[int]) -> dict[str, object]:
    async with AblyRest(ABLY_API_KEY) as ably:
        res = await ably.auth.create_token_request(
            {
                "clientId": user_id,
                "capability": {
                    f"scheduled_recipe:{team_id}": ["subscribe"] for team_id in team_ids
                }
                | {
                    f"cook_checklist:{team_id}:*": ["subscribe"] for team_id in team_ids
                },
            }
        )

        return res.to_dict()


@api_view(["GET"])
@permission_classes([IsAuthenticated])
def ably_detail_view(request: AuthedRequest) -> Response:
    # NOTE: this isn't really scalable if the user has a lot of teams.
    team_ids = list(get_teams(user=request.user).values_list("id", flat=True))
    return Response(
        asyncio.run(get_token(user_id=str(request.user.id), team_ids=team_ids)),
        status=status.HTTP_200_OK,
    )
