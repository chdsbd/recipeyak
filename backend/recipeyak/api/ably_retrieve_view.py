from __future__ import annotations

import asyncio

from ably import AblyRest

from recipeyak.api.base.decorators import endpoint
from recipeyak.api.base.request import AuthedHttpRequest
from recipeyak.api.base.response import JsonResponse
from recipeyak.api.team_update_view import get_teams
from recipeyak.config import ABLY_API_KEY


async def get_token(user_id: str, team_ids: list[int]) -> dict[str, object]:
    async with AblyRest(ABLY_API_KEY) as ably:
        res = await ably.auth.create_token_request(
            {
                "clientId": user_id,
                "capability": {
                    f"scheduled_recipe:{team_id}": ["subscribe"] for team_id in team_ids
                }
                | {
                    f"cook_checklist:{team_id}:*": ["subscribe"] for team_id in team_ids
                },
            }
        )

        return res.to_dict()


@endpoint()
def ably_retrieve_view(request: AuthedHttpRequest) -> JsonResponse:
    # NOTE: this isn't really scalable if the user has a lot of teams.
    team_ids = list(get_teams(user=request.user).values_list("id", flat=True))
    return JsonResponse(
        asyncio.run(get_token(user_id=str(request.user.id), team_ids=team_ids)),
        status=200,
    )
