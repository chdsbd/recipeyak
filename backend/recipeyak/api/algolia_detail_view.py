from __future__ import annotations

import asyncio

from algoliasearch.search_client import SearchClient
from rest_framework import status
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from recipeyak import config
from recipeyak.api.base.request import AuthedRequest
from recipeyak.api.team_update_view import get_teams


async def get_token(team_ids: list[int]) -> str:
    async with SearchClient.create(
        app_id=config.ALGOLIA_APPLICATION_ID, api_key=config.ALGOLIA_ADMIN_API_KEY
    ) as client:
        # https://www.algolia.com/doc/api-reference/api-methods/generate-secured-api-key/
        return client.generate_secured_api_key(
            config.ALGOLIA_SEARCH_ONLY_API_KEY,
            {
                "restrictIndices": "recipes",
                "filters": " OR ".join(f"team_id:{team_id}" for team_id in team_ids),
            },
        )


@api_view(["GET"])
@permission_classes([IsAuthenticated])
def algolia_detail_view(request: AuthedRequest) -> Response:
    # NOTE: this isn't really scalable if the user has a lot of teams.
    team_ids = list(get_teams(user=request.user).values_list("id", flat=True))
    return Response(
        {
            "app_id": config.ALGOLIA_APPLICATION_ID,
            "api_key": asyncio.run(get_token(team_ids=team_ids)),
        },
        status=status.HTTP_200_OK,
    )
