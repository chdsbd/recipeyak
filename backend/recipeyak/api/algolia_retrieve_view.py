from __future__ import annotations

from algoliasearch.search_client import SearchClient
from django.http import JsonResponse

from recipeyak import config
from recipeyak.api.base.decorators import endpoint
from recipeyak.api.base.request import AuthedHttpRequest
from recipeyak.api.team_update_view import get_teams


def get_token(team_ids: list[int]) -> str:
    # https://www.algolia.com/doc/api-reference/api-methods/generate-secured-api-key/
    return SearchClient.generate_secured_api_key(
        config.ALGOLIA_SEARCH_ONLY_API_KEY,
        {
            "restrictIndices": "recipes",
            "filters": " OR ".join(f"team_id:{team_id}" for team_id in team_ids),
        },
    )


@endpoint()
def algolia_retrieve_view(request: AuthedHttpRequest) -> JsonResponse:
    # NOTE: this isn't really scalable if the user has a lot of teams.
    team_ids = list(get_teams(user=request.user).values_list("id", flat=True))
    return JsonResponse(
        {
            "app_id": config.ALGOLIA_APPLICATION_ID,
            "api_key": get_token(team_ids=team_ids),
        },
        status=200,
    )
