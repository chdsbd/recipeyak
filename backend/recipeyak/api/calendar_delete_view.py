from __future__ import annotations

from rest_framework import status
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from recipeyak.api.base.request import AuthedRequest
from recipeyak.api.calendar_list_view import get_scheduled_recipes
from recipeyak.models import get_team
from recipeyak.realtime import publish_calendar_event_deleted


@api_view(["DELETE"])
@permission_classes([IsAuthenticated])
def calendar_delete_view(
    request: AuthedRequest, scheduled_recipe_id: int, team_id: object = ()
) -> Response:
    team_id = get_team(request).id
    get_scheduled_recipes(team_id).filter(id=scheduled_recipe_id).delete()
    publish_calendar_event_deleted(recipe_id=scheduled_recipe_id, team_id=team_id)
    return Response(status=status.HTTP_204_NO_CONTENT)
