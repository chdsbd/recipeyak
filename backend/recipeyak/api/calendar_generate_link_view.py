from django.shortcuts import get_object_or_404
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from recipeyak.api.base.request import AuthedRequest
from recipeyak.api.calendar_list_view import get_cal_settings
from recipeyak.models import Membership, get_random_ical_id, get_team


@api_view(["POST"])
@permission_classes([IsAuthenticated])
def calendar_generate_link_view(request: AuthedRequest, team_id: int = -1) -> Response:
    team_id = get_team(request).id
    membership = get_object_or_404(Membership, team=team_id, user=request.user)
    membership.calendar_secret_key = get_random_ical_id()
    membership.save()

    return Response(get_cal_settings(request=request, team_id=team_id))
