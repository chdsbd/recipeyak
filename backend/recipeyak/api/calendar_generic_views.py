from __future__ import annotations

import logging
from typing import Optional, TypeVar

from django.db import connection
from django.shortcuts import get_object_or_404
from rest_framework import serializers
from rest_framework.decorators import action
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from recipeyak.api.base import viewsets
from recipeyak.api.base.permissions import IsTeamMember
from recipeyak.api.base.request import AuthedRequest
from recipeyak.api.calendar_list_view import (
    ScheduledRecipeSerializer,
    get_cal_settings,
    get_scheduled_recipes,
)
from recipeyak.models import Membership, get_random_ical_id

logger = logging.getLogger(__name__)

T = TypeVar("T")


def unwrap(arg: Optional[T]) -> T:
    assert arg is not None
    return arg


class CalSettingsSerializer(serializers.Serializer):
    syncEnabled = serializers.BooleanField()


class CalendarViewSet(viewsets.ModelViewSet):
    serializer_class = ScheduledRecipeSerializer
    permission_classes = (IsAuthenticated, IsTeamMember)

    # patch method to update the `on` date uses this
    def get_queryset(self):
        return get_scheduled_recipes(self.request, self.kwargs["team_pk"])

    @action(detail=False, methods=["PATCH"], url_path="settings")
    def update_settings(self, request: AuthedRequest, team_pk: str) -> Response:
        serializer = CalSettingsSerializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        sync_enabled = serializer.validated_data["syncEnabled"]

        membership = get_object_or_404(Membership, team=team_pk, user=request.user)
        membership.calendar_sync_enabled = sync_enabled
        membership.save()

        return Response(get_cal_settings(request=request, team_pk=team_pk))

    @action(detail=False, methods=["POST"])
    def generate_link(self, request: AuthedRequest, team_pk: str) -> Response:
        membership = get_object_or_404(Membership, team=team_pk, user=request.user)
        membership.calendar_secret_key = get_random_ical_id()
        membership.save()

        return Response(get_cal_settings(request=request, team_pk=team_pk))

    @action(detail=False, methods=["GET"])
    def next_open(self, request: AuthedRequest, team_pk: str) -> Response:
        with connection.cursor() as cursor:
            weekday = request.query_params["day"]
            now = request.query_params["now"]
            day_number = {
                "Sunday": (0,),
                "Monday": (1,),
                "Tuesday": (2,),
                "Wednesday": (3,),
                "Thursday": (4,),
                "Friday": (5,),
                "Saturday": (6,),
                "Weekday": (1, 2, 3, 4, 5),
                "Weekend": (0, 6),
            }[weekday]

            cursor.execute(
                """
SELECT
  gen_date::date "date"
FROM
  generate_series(%(now)s::date + '1 day'::interval, (%(now)s::date + '3 years'::interval), '1 day') "gen_date"
  LEFT JOIN core_scheduledrecipe ON core_scheduledrecipe."on" = gen_date
    AND team_id = %(team_id)s
WHERE
  date_part('dow', gen_date) in %(day_number)s
  AND id IS NULL
ORDER BY
  gen_date ASC
LIMIT 1;
""",
                {"day_number": day_number, "team_id": team_pk, "now": now},
            )
            (date,) = unwrap(cursor.fetchone())
            return Response({"date": date})
