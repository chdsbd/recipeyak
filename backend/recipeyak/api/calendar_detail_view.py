from __future__ import annotations

from rest_framework.decorators import api_view, permission_classes
from rest_framework.exceptions import MethodNotAllowed
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from recipeyak.api.base.permissions import IsTeamMember
from recipeyak.api.base.request import AuthedRequest
from recipeyak.api.calendar_delete_view import calendar_delete_view
from recipeyak.api.calendar_update_view import calendar_update_view


@api_view(["PATCH", "DELETE"])
@permission_classes([IsAuthenticated, IsTeamMember])
def calendar_detail_view(request: AuthedRequest, team_pk: int, pk: int) -> Response:
    if request.method == "PATCH":
        return calendar_update_view(request, team_pk, pk)
    elif request.method == "DELETE":
        return calendar_delete_view(request, team_pk, pk)
    raise MethodNotAllowed(request.method or "")
