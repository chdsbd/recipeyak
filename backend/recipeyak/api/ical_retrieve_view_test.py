from datetime import date, datetime
from email.utils import parsedate_to_datetime
from urllib.parse import urlparse

import pytest
from django.test.client import Client

from recipeyak.api.ical_retrieve_view import to_ical_time
from recipeyak.models import Recipe, ScheduledRecipe, Team, User, get_random_ical_id
from recipeyak.models.membership import Membership

pytestmark = pytest.mark.django_db


def test_ical_view_with_invalid_id(client: Client, user: User, team: Team) -> None:
    """
    We should return a 404 when a user doesn't pass the correct id for the schedule.
    """
    fake_team_id = get_random_ical_id()
    res = client.get(f"/t/{team.id}/ical/{fake_team_id}/schedule.ics")
    assert res.status_code == 404


def omit_entry_ids(content: str) -> str:
    output = ""
    for i, line in enumerate(content.split("\r\n")):
        if i > 0:
            output += "\r\n"
        if line.startswith("UID"):
            output += "UID:<id-removed>"
        else:
            output += line
    return output


def test_ical_ids_consistent(
    client: Client, user: User, recipe: Recipe, team: Team
) -> None:
    """
    Regression test to ensure the ids for the items aren't changing on each
    request.
    """
    ScheduledRecipe.objects.create(recipe=recipe, team=team, on=date(1976, 7, 6))
    ScheduledRecipe.objects.create(recipe=recipe, team=team, on=date(1976, 7, 7))
    ScheduledRecipe.objects.create(recipe=recipe, team=team, on=date(1976, 7, 10))
    membership = Membership.objects.filter(user=user).get(team=team)
    membership.calendar_sync_enabled = True
    membership.save()
    ical_id = membership.calendar_secret_key
    url = f"/t/{team.id}/ical/{ical_id}/schedule.ics"
    res = client.get(url)
    assert res.status_code == 200

    res_second = client.get(url)
    assert res_second.status_code == 200

    assert (
        res.content == res_second.content
    ), "Ensure we don't have ids being regenerated and changing on each request"


def test_ical_view_with_correct_id(
    client: Client, user: User, recipe: Recipe, team: Team
) -> None:
    """
    When the client passes in the correct id we should return the ical data.
    We omit the ids to make the snapshot style test work with the random
    uuids used by the calendar entries.
    """

    scheduled_a = ScheduledRecipe.objects.create(
        recipe=recipe,
        team=team,
        on=date(1976, 7, 6),
    )
    scheduled_b = ScheduledRecipe.objects.create(
        recipe=recipe,
        team=team,
        on=date(1976, 7, 7),
    )
    scheduled_c = ScheduledRecipe.objects.create(
        recipe=recipe,
        team=team,
        on=date(1976, 7, 10),
    )

    membership = Membership.objects.filter(user=user).get(team=team)
    membership.calendar_sync_enabled = True
    membership.save()
    ical_id = membership.calendar_secret_key
    url = f"/t/{team.id}/ical/{ical_id}/schedule.ics"
    res = client.get(url)
    assert res.status_code == 200
    assert res["Content-Type"] == "text/calendar"
    assert isinstance(parsedate_to_datetime(res["Last-Modified"]), datetime)

    assert omit_entry_ids(res.content.decode()) == (
        "BEGIN:VCALENDAR\r\n"
        "VERSION:2.0\r\n"
        "PRODID:-//Recipe Yak//Schedule//EN\r\n"
        "CALSCALE:GREGORIAN\r\n"
        "X-WR-CALDESC:Recipe Yak Schedule for Team Recipe Yak Team\r\n"
        "X-WR-CALNAME:Scheduled Recipes\r\n"
        "BEGIN:VEVENT\r\n"
        "SUMMARY:Recipe name\r\n"
        "DTSTART;VALUE=DATE:19760706\r\n"
        "DTEND;VALUE=DATE:19760707\r\n"
        f"DTSTAMP:{to_ical_time(scheduled_a.created)}\r\n"
        "UID:<id-removed>\r\n"
        f"CREATED:{to_ical_time(scheduled_a.created)}\r\n"
        "DESCRIPTION:Takes about 1 hour\r\n"
        f"LAST-MODIFIED:{to_ical_time(scheduled_a.created)}\r\n"
        "TRANSP:TRANSPARENT\r\n"
        f"URL:https://recipeyak.com/recipes/{scheduled_a.recipe_id}-recipe-name\r\n"
        "END:VEVENT\r\n"
        "BEGIN:VEVENT\r\n"
        "SUMMARY:Recipe name\r\n"
        "DTSTART;VALUE=DATE:19760707\r\n"
        "DTEND;VALUE=DATE:19760708\r\n"
        f"DTSTAMP:{to_ical_time(scheduled_b.created)}\r\n"
        "UID:<id-removed>\r\n"
        f"CREATED:{to_ical_time(scheduled_b.created)}\r\n"
        "DESCRIPTION:Takes about 1 hour\r\n"
        f"LAST-MODIFIED:{to_ical_time(scheduled_b.created)}\r\n"
        "TRANSP:TRANSPARENT\r\n"
        f"URL:https://recipeyak.com/recipes/{scheduled_b.recipe_id}-recipe-name\r\n"
        "END:VEVENT\r\n"
        "BEGIN:VEVENT\r\n"
        "SUMMARY:Recipe name\r\n"
        "DTSTART;VALUE=DATE:19760710\r\n"
        "DTEND;VALUE=DATE:19760711\r\n"
        f"DTSTAMP:{to_ical_time(scheduled_c.created)}\r\n"
        "UID:<id-removed>\r\n"
        f"CREATED:{to_ical_time(scheduled_c.created)}\r\n"
        "DESCRIPTION:Takes about 1 hour\r\n"
        f"LAST-MODIFIED:{to_ical_time(scheduled_c.created)}\r\n"
        "TRANSP:TRANSPARENT\r\n"
        f"URL:https://recipeyak.com/recipes/{scheduled_c.recipe_id}-recipe-name\r\n"
        "END:VEVENT\r\n"
        "END:VCALENDAR\r\n"
    )


def test_get_ical_view_works_with_accept_encoding(
    client: Client, user: User, recipe: Recipe, team: Team
) -> None:
    """
    Clicking an .ics link on iOS will trigger iOS to make a GET request to
    the URL with accept header set to `text/calendar`.

    When the view is setup using DRF this fails and returns a 406, which
    prevents the Calendar app from using the "subscribed" calendar.
    """
    membership = Membership.objects.filter(user=user).get(team=team)
    membership.calendar_sync_enabled = True
    membership.save()
    ical_id = membership.calendar_secret_key
    url = f"/t/{team.id}/ical/{ical_id}/schedule.ics"
    res = client.get(url, HTTP_ACCEPT="text/calendar")
    assert res.status_code == 200


def test_head_ical(client: Client, user: User, recipe: Recipe, team: Team) -> None:
    """
    HEAD should work
    """
    membership = Membership.objects.filter(user=user).get(team=team)
    membership.calendar_sync_enabled = True
    membership.save()
    ical_id = membership.calendar_secret_key
    url = f"/t/{team.id}/ical/{ical_id}/schedule.ics"
    res = client.head(url, HTTP_ACCEPT="text/calendar")
    assert res.status_code == 200


def test_get_ical_view_with_user_specific_id(
    client: Client, user: User, recipe: Recipe, team: Team
) -> None:
    """
    Ensure the user specific secret key works.
    """
    membership = Membership.objects.filter(user=user).get(team=team)
    membership.calendar_sync_enabled = True
    membership.save()
    url = f"/t/{team.id}/ical/{membership.calendar_secret_key}/schedule.ics"
    res = client.get(url, HTTP_ACCEPT="text/calendar")
    assert res.status_code == 200


def test_get_ical_view_with_schedule_endpoint(
    client: Client, user: User, recipe: Recipe, team: Team
) -> None:
    """
    Ensure the url from the schedule endpoint is valid.
    """
    url = url = f"/api/v1/t/{team.pk}/calendar/"
    client.force_login(user)

    res = client.get(url, {"start": date(1976, 1, 1), "end": date(1977, 1, 1), "v2": 1})
    assert res.status_code == 200
    calendar_link = res.json()["settings"]["calendarLink"]

    client.logout()
    relative_url = urlparse(calendar_link).path

    membership = user.membership_set.get(team=team)
    membership.calendar_sync_enabled = True
    membership.save()

    res = client.get(relative_url, HTTP_ACCEPT="text/calendar")
    assert (
        res.status_code == 200
    ), "API returned a valid URL for retrieving calendar info."


def test_get_ical_view_404_when_disabled(
    client: Client, user: User, recipe: Recipe, team: Team
) -> None:
    """
    Url should be disabled when calendar_sync_enabled is false.
    """
    membership = Membership.objects.filter(user=user).get(team=team)
    membership.calendar_sync_enabled = True
    membership.save()
    url = f"/t/{team.id}/ical/{membership.calendar_secret_key}/schedule.ics"
    res = client.get(url, HTTP_ACCEPT="text/calendar")
    assert res.status_code == 200

    membership.calendar_sync_enabled = False
    membership.save()
    res = client.get(url, HTTP_ACCEPT="text/calendar")
    assert res.status_code == 404
