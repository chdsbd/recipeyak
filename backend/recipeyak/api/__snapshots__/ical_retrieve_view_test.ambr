# name: test_ical_view_with_correct_id
  '''
  BEGIN:VCALENDAR
  PRODID:-//Recipe Yak//Schedule//EN
  CALSCALE:GREGORIAN
  VERSION:2.0
  X-WR-CALNAME:Scheduled Recipes
  X-WR-CALDESC:Recipe Yak Schedule for Team Recipe Yak Team
  BEGIN:VEVENT
  UID:<id-removed>
  DTSTART;VALUE=DATE:19760706
  DTEND;VALUE=DATE:19760707
  SUMMARY:Recipe name
  DESCRIPTION:Takes about 1 hour
  URL:https://recipeyak.com/recipes/<id-removed>
  TRANSP:TRANSPARENT
  CREATED:<timestamp-removed>
  DTSTAMP:<timestamp-removed>
  LAST-MODIFIED:<timestamp-removed>
  END:VEVENT
  BEGIN:VEVENT
  UID:<id-removed>
  DTSTART;VALUE=DATE:19760707
  DTEND;VALUE=DATE:19760708
  SUMMARY:Recipe name
  DESCRIPTION:Takes about 1 hour
  URL:https://recipeyak.com/recipes/<id-removed>
  TRANSP:TRANSPARENT
  CREATED:<timestamp-removed>
  DTSTAMP:<timestamp-removed>
  LAST-MODIFIED:<timestamp-removed>
  END:VEVENT
  BEGIN:VEVENT
  UID:<id-removed>
  DTSTART;VALUE=DATE:19760710
  DTEND;VALUE=DATE:19760711
  SUMMARY:Recipe name
  DESCRIPTION:Takes about 1 hour
  URL:https://recipeyak.com/recipes/<id-removed>
  TRANSP:TRANSPARENT
  CREATED:<timestamp-removed>
  DTSTAMP:<timestamp-removed>
  LAST-MODIFIED:<timestamp-removed>
  END:VEVENT
  END:VCALENDAR
  
  '''
# ---
